### PR TITLE
modify base test class analyzer, allow to extend from ft base test class

### PR DESCRIPTION
**Pull request by frank-dong-ms**
We have separate base test class for all functional tests: FunctionalTestBaseClass
The reason for that is FT is also runned on NightlyBuild pipeline which not take project dependency directly but take ML.NET related Nuget package as dependency and we don't want FT to take dependency from TestFramework.

### CODE REVIEW

File: tools-local/Microsoft.ML.InternalCodeAnalyzer/BaseTestClassAnalyzer.cs [hunk 1]
```diff
@@ -51,13 +51,15 @@ private sealed class AnalyzerImpl
...
             private readonly INamedTypeSymbol _baseTestClass;
+            private readonly INamedTypeSymbol _FTbaseTestClass;
             private readonly ConcurrentDictionary<INamedTypeSymbol, bool> _knownTestAttributes = new ConcurrentDictionary<INamedTypeSymbol, bool>();
 
             public AnalyzerImpl(Compilation compilation, INamedTypeSymbol factAttribute)
             {
                 _compilation = compilation;
                 _factAttribute = factAttribute;
                 _baseTestClass = _compilation.GetTypeByMetadataName("Microsoft.ML.TestFramework.BaseTestClass");
+                _FTbaseTestClass = _compilation.GetTypeByMetadataName("Microsoft.ML.Functional.Tests.FunctionalTestBaseClass");
```
Comment (harishsk) [inline tools-local/Microsoft.ML.InternalCodeAnalyzer/BaseTestClassAnalyzer.cs]: Why do we have two base test classes? I know you have investigated this before. Can you please help me understand why we can't have a single BaseTestClass and have FunctionalTestBaseClass derive from BaseTestClass? (From what I understand, the only difference for functional tests is that they only use public API)

Comment (frank-dong-ms) [reply tools-local/Microsoft.ML.InternalCodeAnalyzer/BaseTestClassAnalyzer.cs]: Yes, please reference previous PR for this: [https://github.com/dotnet/machinelearning/pull/4346](https://github.com/dotnet/machinelearning/pull/4346)

Basically, we want FT to be able to dependent on public nuget package so we need to break FT from any direct or indirect project reference. TestFramework project itself has lots of project dependency and also been marked as BestFriend from some projects from source folder which allows TestFramework peoject to access some non public code. So, we extracted TestFrameworkCommon test project which only have public dependency. Previously FTs are inherited from BaseTestBaseline which has lots of BestFriend thing and can't be moved to TestFrameworkCommon project so after discuss we decide to write one base class for FT only with only public dependency.
So in short we have separate base class for FT because we don't want to take project dependency from our source code to let FT able to run in NightlyBuild

---
In reply to: [376581212](https://github.com/dotnet/machinelearning/pull/4810#discussion_r376581212) [](ancestors = 376581212)

Comment (harishsk) [reply tools-local/Microsoft.ML.InternalCodeAnalyzer/BaseTestClassAnalyzer.cs]: You are right! That sounds good. In that case, please make sure the functionality between the two base classes are in sync.  I recently added some code to extract the LogMessageKind attribute from the test and store it in the base class. I also changed the derived class to append to the logs correctly. Can you please make sure that the important parts from this PR are integrated into the the FunctionalTestBaseClass: [https://github.com/dotnet/machinelearning/pull/4710](https://github.com/dotnet/machinelearning/pull/4710)

---
In reply to: [376616140](https://github.com/dotnet/machinelearning/pull/4810#discussion_r376616140) [](ancestors = 376616140,376581212)

Comment (harishsk) [reply tools-local/Microsoft.ML.InternalCodeAnalyzer/BaseTestClassAnalyzer.cs]: The above mentioned work need not be part of this PR. I am approving this. Please use a different PR for that work.

Comment (harishsk) [reply tools-local/Microsoft.ML.InternalCodeAnalyzer/BaseTestClassAnalyzer.cs]: I am approving this. You can do the above mentioned work in a different PR.

Comment (frank-dong-ms) [reply tools-local/Microsoft.ML.InternalCodeAnalyzer/BaseTestClassAnalyzer.cs]: Sure, I will open a separate PR if we need certain fix

---
In reply to: [376622761](https://github.com/dotnet/machinelearning/pull/4810#discussion_r376622761) [](ancestors = 376622761)
